### PR TITLE
Feature/col 135/device health rule design

### PR DESCRIPTION
**Pull request by ddarshan-juniper**
Contains the following changes:
1. Addition of <kpi-field>-anomaly which calls the "jaiml-anomaly-detection" module to get anomaly status.
2. Inclusion of the above new fields into triggers adhering to the rule:  static rule > anomaly status .
3. As a temporary solution any value less than or equal to will be stated "NOT ANOMALOUS"

TODO: 
1. Logic to handle error conditions separately (eg: when status code is -1, -2 , -3 etc..)

### CODE REVIEW

File: juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule [hunk 1]
```diff
@@ -254,10 +300,29 @@ healthbot {
                         }
                     }
                 }
+                *
+                * Sets color to yellow when fpc-cpu-utilization utilization is below low threshold ($low-threshold)
+                * and the fpc-cpu-utilization value is anomalous.
+                */ 
+                term is-fpc-cpu-utilization-anomalous {
+                    when {
+                        less-than-or-equal-to "$fpc-cpu-utilization" "$low-threshold";
```
Comment (gaddevijay) [inline juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule:309]: Is this required?

File: juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule [hunk 2]
```diff
@@ -254,10 +300,29 @@ healthbot {
...
+                        }
+                    }
+                }
                 /*
-                 * Defaults color to green.
+                 * Defaults color to green when fpc-cpu-utilization is below low-threshold and not anomalous
                  */				
                 term fpc-cpu-utilization-normal {
+                    when {
+                        less-than-or-equal-to "$fpc-cpu-utilization-anomaly" 0;
```
Comment (gaddevijay) [inline juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule:324]: This also is not required right?

File: juniper_official/chassis/check-fan-state-rpm.rule [hunk 3]
```diff
@@ -165,10 +174,29 @@ healthbot {
...
+                 */ 
+                term is-rpm-percent-anomalous {
+                    when {
+                        less-than-or-equal-to "$rpm-percent" "$low-threshold";
+                        equal-to "$rpm-percent-anomaly" 1;
+                    }
+                    then {
+                        status {
+                            color yellow;
+                            message "$fan-name RPM percent($rpm-percent%) is not between the acceptable range of ($rpm-percent-anomaly-lower and $rpm-percent-anomaly-upper)";
```
Comment (gaddevijay) [inline juniper_official/chassis/check-fan-state-rpm.rule:188]: Seems like return of JAIML is upper-boundary and lower-boundary. Then field will be like <field>-upper-boundary instead for <field>-upper

**Review by gaddevijay — CHANGES_REQUESTED** [review]

File: juniper_official/chassis/check-fan-state-rpm.rule [hunk 4]
```diff
@@ -165,10 +174,29 @@ healthbot {
                     }
                 }
                 /*
-                 * Sets color to green when RPM is less than 50%.
+                 * Sets color to yellow when RPM percent is below low threshold ($low-threshold)
+                 * and the rpm-percent value is anomalous.
+                 */ 
+                term is-rpm-percent-anomalous {
+                    when {
+                        less-than-or-equal-to "$rpm-percent" "$low-threshold";
```
[the remark this answers is not in the document]
Comment (ddarshan-juniper) [reply juniper_official/chassis/check-fan-state-rpm.rule]: removed redundant checks

File: juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule [hunk 5]
```diff
@@ -211,9 +238,28 @@ healthbot {
...
+                            message "$fpc buffer memory utilization($fpc-memory-buffer) is not between the acceptable range of ($fpc-memory-buffer-anomaly-lower and $fpc-memory-buffer-anomaly-upper)";
+                        }
+                    }
+                }
+                /*
+                 * Defaults color to green when its less than low-threshold and fpc-memory-buffer value is not anomalous
                  */				
                 term fpc-buffer-memory-utilization-normal {
+                    when {
+                        less-than-or-equal-to "$fpc-memory-buffer-anomaly" 0;
```
[the remark this answers is not in the document]
Comment (ddarshan-juniper) [reply juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule]: removed redundant checks

File: juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule (hunk 2 above)
Comment (ddarshan-juniper) [reply juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule]: removed redundant checks

File: juniper_official/chassis/check-fan-state-rpm.rule (hunk 3 above)
Comment (ddarshan-juniper) [reply juniper_official/chassis/check-fan-state-rpm.rule]: added the missed "-boundary" tag

File: juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule [hunk 6]
```diff
@@ -356,16 +413,16 @@ healthbot {
                 description "Sets health based on FPC Temperature";
                 frequency 2offset;
                 /*
-                 * Sets color to green when temperature is less-than-or-equal-to temp-low-threshold.
+                 * Sets color to red when temperature is-greater-than-or-equal-to temp-high-threshold .
                  */				
-                term is-temperature-normal {
```
Comment (gaddevijay) [inline juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule:361]: Retain normal check first for the triggers which are already adapted to this way

**Review by gaddevijay — CHANGES_REQUESTED** [review]

File: juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule (hunk 6 above)
Comment (ddarshan-juniper) [reply juniper_official/linecard/check-fpc-cpu-memory-state-temp.rule:361]: Have changed the sequence to original format

File: juniper_official/chassis/check-routing-engine-temperature.rule [hunk 7]
```diff
@@ -82,36 +82,56 @@ healthbot {
...
                 synopsis "RE temperature KPI";
                 description "Sets health based increase in RE temperature";
                 frequency 1offset;
+                
                 /*
-                 * Sets color to green when RE temperature
-                 * re-temperature is below low threshold ($low-threshold).
+                 * Defaults color to red and notify anomaly.
                  */ 
-                term is-re-temperature-normal {
```
[the remark this answers is not in the document]
Comment (ddarshan-juniper) [reply juniper_official/chassis/check-routing-engine-temperature.rule:96]: Have changed the sequence to original format

**Pull request merged**